# Севастьянов Даниил 10И1


In [ ]:
from google.colab import files
uploaded = files.upload()

Saving diabetes.csv to diabetes.csv


In [ ]:
import pandas as pd
df = pd.read_csv("diabetes.csv")
df

Pregnancies  Glucose  ...  Age  Outcome
0              6      148  ...   50        1
1              1       85  ...   31        0
2              8      183  ...   32        1
3              1       89  ...   21        0
4              0      137  ...   33        1
..           ...      ...  ...  ...      ...
763           10      101  ...   63        0
764            2      122  ...   27        0
765            5      121  ...   30        0
766            1      126  ...   47        1
767            1       93  ...   23        0

[768 rows x 9 columns]

In [ ]:
print(df.shape)
print('###########################################')
print(df.isna().sum())
print('#######################################')
print(df.info())

(768, 9)
###########################################
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64
#######################################
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                 

In [ ]:
Делаем масштабирование признаков: нормализацию и стандартизацию

In [ ]:
X = df.drop(['Outcome'], axis = 1)
y = df['Outcome']
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_st = scaler.fit_transform(X)
print(X_st)
print('##############'*5)
scaler = MinMaxScaler()
X_norm = scaler.fit_transform(X)
X_norm

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]
######################################################################


array([[0.35294118, 0.74371859, 0.59016393, ..., 0.50074516, 0.23441503,
        0.48333333],
       [0.05882353, 0.42713568, 0.54098361, ..., 0.39642325, 0.11656704,
        0.16666667],
       [0.47058824, 0.91959799, 0.52459016, ..., 0.34724292, 0.25362938,
        0.18333333],
       ...,
       [0.29411765, 0.6080402 , 0.59016393, ..., 0.390462  , 0.07130658,
        0.15      ],
       [0.05882353, 0.63316583, 0.49180328, ..., 0.4485842 , 0.11571307,
        0.43333333],
       [0.05882353, 0.46733668, 0.57377049, ..., 0.45305514, 0.10119556,
        0.03333333]])

knn со стандартизированными данными

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X_st, y, test_size=1/3, random_state=42)
knn = KNeighborsClassifier(5)

knn.fit(X_train,y_train)
y_pred = knn.predict(X_test)
confusion_matrix(y_test,y_pred)

print("Accuracy:", accuracy_score(y_test,y_pred))
print("Recall:", recall_score(y_test,y_pred))
print("Precision:", precision_score(y_test,y_pred))

Accuracy: 0.6875
Recall: 0.5
Precision: 0.5375


knn с нормализованными данными

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)
knn = KNeighborsClassifier(5)

knn.fit(X_train,y_train)
y_pred = knn.predict(X_test)
confusion_matrix(y_test,y_pred)

print("Accuracy:", accuracy_score(y_test,y_pred))
print("Recall:", recall_score(y_test,y_pred))
print("Precision:", precision_score(y_test,y_pred))

Accuracy: 0.6796875
Recall: 0.5116279069767442
Precision: 0.5238095238095238


Логистическая регрессия со стандартизированными данными

In [ ]:
df1=df
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_st,y, test_size=1/3)

model = LogisticRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(model.score(X_test, y_test))
print("Accuracy:", accuracy_score(y_test,y_pred))
print("Recall:", recall_score(y_test,y_pred))
print("Precision:", precision_score(y_test,y_pred))

0.80859375
Accuracy: 0.80859375
Recall: 0.6309523809523809
Precision: 0.7464788732394366


Логистическая регрессия с нормализованными данными

In [ ]:
df1=df
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_norm,y, test_size=1/3)

model = LogisticRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(model.score(X_test, y_test))
print("Accuracy:", accuracy_score(y_test,y_pred))
print("Recall:", recall_score(y_test,y_pred))
print("Precision:", precision_score(y_test,y_pred))

0.734375
Accuracy: 0.734375
Recall: 0.44329896907216493
Precision: 0.7543859649122807


Дерево решений с нормализованными:

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)
tree = DecisionTreeClassifier(criterion='gini', max_depth=4, random_state=1)

tree.fit(X_train, y_train)
y_pred_tree = tree.predict(X_test)

print("Accuracy:", accuracy_score(y_test,y_pred_tree))
print("Recall:", recall_score(y_test,y_pred_tree))
print("Precision:", precision_score(y_test,y_pred_tree))

Accuracy: 0.72265625
Recall: 0.7209302325581395
Precision: 0.5688073394495413


Дерево решений со стандартизированными:

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X_st, y, test_size=1/3, random_state=42)
tree = DecisionTreeClassifier(criterion='gini', max_depth=4, random_state=1)

tree.fit(X_train, y_train)
y_pred_tree = tree.predict(X_test)

print("Accuracy:", accuracy_score(y_test,y_pred_tree))
print("Recall:", recall_score(y_test,y_pred_tree))
print("Precision:", precision_score(y_test,y_pred_tree))

Accuracy: 0.72265625
Recall: 0.7209302325581395
Precision: 0.5688073394495413


In [ ]:
оказалось, что точность одинакова при нормализованных и стандортизированных

поменяем глубину в decisiontreeclassifier

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)
tree = DecisionTreeClassifier(criterion='gini', max_depth=15, random_state=1)

tree.fit(X_train, y_train)
y_pred_tree = tree.predict(X_test)

print("Accuracy:", accuracy_score(y_test,y_pred_tree))
print("Recall:", recall_score(y_test,y_pred_tree))
print("Precision:", precision_score(y_test,y_pred_tree))

Accuracy: 0.70703125
Recall: 0.6627906976744186
Precision: 0.5533980582524272


In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)
tree = DecisionTreeClassifier(criterion='gini', max_depth=7, random_state=1)

tree.fit(X_train, y_train)
y_pred_tree = tree.predict(X_test)

print("Accuracy:", accuracy_score(y_test,y_pred_tree))
print("Recall:", recall_score(y_test,y_pred_tree))
print("Precision:", precision_score(y_test,y_pred_tree))

Accuracy: 0.7421875
Recall: 0.6511627906976745
Precision: 0.6086956521739131


In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

X_train,X_test,y_train,y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)
tree = DecisionTreeClassifier(criterion='gini', max_depth=7, random_state=1)

tree.fit(X_train, y_train)
y_pred_tree = tree.predict(X_test)

print("Accuracy:", accuracy_score(y_test,y_pred_tree))
print("Recall:", recall_score(y_test,y_pred_tree))
print("Precision:", precision_score(y_test,y_pred_tree))

Accuracy: 0.71875
Recall: 0.313953488372093
Precision: 0.675


Видно что наибольшее accuracy при max_depth=7 и равно примерно 74%

построим древо решений с крос валидацией при max_depth=7

In [ ]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold

kf = KFold(n_splits=5)
 
tree = DecisionTreeClassifier(criterion='gini', max_depth=7, random_state=1)
cvs = cross_val_score(estimator=tree, X=X, y=y, cv=kf, scoring='accuracy')

print(cvs)
print(cvs.mean())

[0.71428571 0.62987013 0.72727273 0.83006536 0.73856209]
0.7280112044817927


In [ ]:
видно что точность упала до 72%, хотя максимальное значение достигало 83%

теперь сделаем логистическую решрессию с крос валидацией

In [ ]:
from sklearn.model_selection import cross_val_score
cvs = cross_val_score(model, X_norm, y, cv=10)
print(cvs)
print(cvs.mean())

[0.75324675 0.74025974 0.79220779 0.7012987  0.74025974 0.80519481
 0.75324675 0.79220779 0.72368421 0.84210526]
0.7643711551606288


In [ ]:
from sklearn.model_selection import cross_val_score
cvs = cross_val_score(model, X_st, y, cv=10)
print(cvs)
print(cvs.mean())

[0.72727273 0.77922078 0.80519481 0.71428571 0.75324675 0.75324675
 0.80519481 0.80519481 0.75       0.82894737]
0.7721804511278195


В итоге, наивысших результатов(77%) достигла логистическая регрессия со стандартизированными данными и крос валидацией.